graphql_compiler/schema_transformation/split_query.py

```diff
@@ -34,6 +46,392 @@ def __init__(self, query_ast):
+
+    Each cross-schema vertex field will have its branch removed from the output AST, and made
+    into its own separate query AST. The parent and child property fields used in
+    the stitch will be added to the parent and child ASTs. @output directives will be added
+    to the parent and child property fields, if not already present.
+
+    The function will be called recursively on each intra-schema vertex field.
+
+    Args:
+        query_node: SubQueryNode, whose list of child query connections may be modified to
+                    include new children
+        ast: Field, InlineFragment, or OperationDefinition, the AST that we are trying to split
+             into child components. It is not modified by this function
+        type_info: TypeInfo, used to get information about the types of fields while traversing
+                   the query ast
+        edge_to_stitch_fields: Dict[Tuple(str, str), Tuple(str, str)], mapping
+                               (type name, vertex field name) to
+                               (source field name, sink field name) used in the @stitch directive
+                               for each cross schema edge
+        intermediate_out_name_assigner: IntermediateOutNameAssigner, object used to generate
+                                        and keep track of names of newly created @output
+                                        directives
+    """
+    # Split selections into three kinds: property fields, cross schema vertex fields, normal
+    # vertex fields/type coercions
+    type_info.enter(ast.selection_set)
+    parent_type_name = type_info.get_parent_type().name
+
+    property_fields_map, vertex_fields = _split_selections_property_and_vertex(
+        ast.selection_set.selections
+    )
+    intra_schema_fields, cross_schema_fields = _split_vertex_fields_intra_and_cross_schema(
+        vertex_fields, parent_type_name, edge_to_stitch_fields
+    )
+
+    made_changes = False
+    if len(cross_schema_fields) > 0:
+        made_changes = True
+
+    # First, process cross schema fields
+    for cross_schema_field in cross_schema_fields:
+        type_info.enter(cross_schema_field)
+
+        if type_info.get_type() is not None:
+            child_type_name = strip_non_null_and_list_from_type(type_info.get_type()).name
+        else:
+            raise AssertionError(u'')
+
+        parent_field_name, child_field_name = edge_to_stitch_fields[
+            (parent_type_name, cross_schema_field.name.value)
+        ]
+        _process_cross_schema_field(
+            query_node, cross_schema_field, property_fields_map, child_type_name,
+            parent_field_name, child_field_name, intermediate_out_name_assigner
+        )
+        type_info.leave(cross_schema_field)
+
+    # Then, process intra schema edges by recursing on them
+    new_intra_schema_fields = []
+    for intra_schema_field in intra_schema_fields:
+        type_info.enter(intra_schema_field)
```

Comment:
I probably won't have time to get to this PR, unfortunately. For now I've refactored the code a bit, so that it's hopefully easier to see the pairs of enter/leave.